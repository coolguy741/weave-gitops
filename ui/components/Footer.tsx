import * as React from "react";
import styled from "styled-components";
import p from "../../package.json";
import { useVersion } from "../hooks/version";
import { GetVersionResponse } from "../lib/api/core/core.pb";
import Flex from "./Flex";
import Link from "./Link";
import Spacer from "./Spacer";
import Text from "./Text";
import Version, { VersionProps } from "./Version";

type Props = {
  className?: string;
};

const RightFoot = styled(Flex)`
  padding-right: ${(props) => props.theme.spacing.medium};
`;

const LeftFoot = styled(Flex)``;

const REPO_URL = "https://github.com/weaveworks/weave-gitops";

function Footer({ className }: Props) {
  const { data, isLoading } = useVersion();
  const versionData = data || ({} as GetVersionResponse);

  const shouldDisplayApiVersion =
    !isLoading &&
    (versionData.semver || "").replace(/^v+/, "") !== p.version &&
    versionData.branch &&
    versionData.commit;

  const wegoVersionText = shouldDisplayApiVersion
    ? `${versionData.branch}-${versionData.commit}`
    : `v${p.version}`;
  const wegoVersionHref = shouldDisplayApiVersion
    ? `${REPO_URL}/commit/${versionData.commit}`
    : `${REPO_URL}/releases/tag/v${p.version}`;

  const versions: VersionProps[] = !isLoading
    ? [
        {
          productName: "Kubernetes",
          versionText: versionData.kubeVersion,
        },
        {
          productName: "Flux",
          versionText: versionData.fluxVersion,
        },
        {
          productName: "Weave GitOps",
          versionText: wegoVersionText,
          versionHref: wegoVersionHref,
        },
      ]
    : [];

  return (
    <Flex as="footer" wide between className={className} role="footer">
      <LeftFoot>
        <Text noWrap>Need help? Contact us at</Text>
        <Spacer padding="xxs" />
        <Link newTab href="mailto:support@weave.works">
          support@weave.works
        </Link>
      </LeftFoot>
      <RightFoot>
        {!isLoading &&
          versions.map((version, index) => (
            <React.Fragment key={`version ${index}`}>
              <Spacer padding="xxs" />
              <Version {...version} />
            </React.Fragment>
          ))}
        <Spacer padding="xxs" />
        <Text noWrap>Â© 2022 Weaveworks</Text>
      </RightFoot>
    </Flex>
  );
}

export default styled(Footer).attrs({ className: Footer.name })`
  color: ${(props) => props.theme.colors.neutral30};
`;
