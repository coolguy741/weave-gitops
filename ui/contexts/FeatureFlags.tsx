import * as React from "react";

export type FeatureFlagsContext = {
  authFlag: boolean | null;
  clusterUserFlag: boolean | null;
  OIDCFlag: boolean | null;
};

export const FeatureFlags =
  React.createContext<FeatureFlagsContext | null>(null);

export default function FeatureFlagsContextProvider({ children }) {
  const [authFlag, setAuthFlag] = React.useState<boolean>(null);
  const [clusterUserFlag, setClusterUserFlag] = React.useState<boolean>(null);
  const [OIDCFlag, setOIDCFlag] = React.useState<boolean>(null);

  const getAuthFlag = React.useCallback(() => {
    fetch("/v1/featureflags")
      .then((response) => response.json())
      .then((data) => {
        setAuthFlag(data.flags.WEAVE_GITOPS_AUTH_ENABLED === "true");
        setClusterUserFlag(data.flags.CLUSTER_USER_AUTH === "true");
        setOIDCFlag(data.flags.OIDC_AUTH === "true");
      })
      .catch((err) => console.log(err));
  }, []);

  React.useEffect(() => {
    getAuthFlag();
  }, [getAuthFlag]);

  // Loading...
  if (authFlag === null) {
    return null;
  }

  return (
    <FeatureFlags.Provider
      value={{
        authFlag,
        clusterUserFlag,
        OIDCFlag,
      }}
    >
      {children}
    </FeatureFlags.Provider>
  );
}
