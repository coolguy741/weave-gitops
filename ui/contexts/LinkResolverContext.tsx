import * as React from "react";

export type LinkResolver = (str: string, params?: any) => string;

interface Props {
  resolver: LinkResolver;
  children: any;
}

const LinkResolverContext = React.createContext<LinkResolver>(null);

export function LinkResolverProvider({ resolver, children }: Props) {
  return (
    <LinkResolverContext.Provider value={resolver}>
      {children}
    </LinkResolverContext.Provider>
  );
}

export function useLinkResolver() {
  return React.useContext(LinkResolverContext);
}
