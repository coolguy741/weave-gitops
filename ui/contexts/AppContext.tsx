import * as React from "react";
import { useHistory } from "react-router-dom";
import { Applications } from "../lib/api/applications/applications.pb";
import { formatURL } from "../lib/nav";
import { FluxObject, FluxObjectNode } from "../lib/objects";
import {
  clearCallbackState,
  getCallbackState,
  getProviderToken,
  storeCallbackState,
  storeProviderToken,
} from "../lib/storage";
import { PageRoute, V2Routes } from "../lib/types";
import { notifySuccess } from "../lib/utils";

type AppState = {
  error: null | { fatal: boolean; message: string; detail?: string };
  nodeYaml: FluxObjectNode | null;
};

type AppSettings = {
  renderFooter: boolean;
};

export type LinkResolver = (incoming: string) => string;

export function defaultLinkResolver(incoming: string): string {
  return incoming;
}

export type AppContextType = {
  applicationsClient: typeof Applications;
  userConfigRepoName: string;
  doAsyncError: (message: string, detail: string) => void;
  clearAsyncError: () => void;
  setNodeYaml: (obj: FluxObject | FluxObjectNode) => void;
  appState: AppState;
  settings: AppSettings;
  linkResolver: LinkResolver;
  getProviderToken: typeof getProviderToken;
  storeProviderToken: typeof storeProviderToken;
  getCallbackState: typeof getCallbackState;
  storeCallbackState: typeof storeCallbackState;
  clearCallbackState: typeof clearCallbackState;
  navigate: {
    internal: (page: PageRoute | V2Routes, query?: any) => void;
    external: (url: string) => void;
  };
  notifySuccess: typeof notifySuccess;
  request: typeof window.fetch;
};

export const AppContext = React.createContext<AppContextType>(
  null as AppContextType
);

export interface AppProps {
  applicationsClient?: typeof Applications;
  linkResolver?: LinkResolver;
  children?: any;
  renderFooter?: boolean;
  notifySuccess?: typeof notifySuccess;
}

export default function AppContextProvider({
  applicationsClient,
  ...props
}: AppProps) {
  const history = useHistory();
  const [appState, setAppState] = React.useState({
    error: null,
    nodeYaml: null,
  });

  const clearAsyncError = () => {
    setAppState({
      ...appState,
      error: null,
    });
  };

  React.useEffect(() => {
    // clear the error state on navigation
    clearAsyncError();
  }, [window.location]);

  const doAsyncError = (message: string, detail: string) => {
    console.error(message);
    setAppState({
      ...appState,
      error: { message, detail },
    });
  };

  const setNodeYaml = (obj: FluxObject | FluxObjectNode) => {
    if (obj) setAppState({ ...appState, nodeYaml: obj });
    else setAppState({ ...appState, nodeYaml: null });
  };

  const value: AppContextType = {
    applicationsClient,
    userConfigRepoName: "wego-github-jlw-config-repo",
    doAsyncError,
    clearAsyncError,
    setNodeYaml,
    appState,
    linkResolver: props.linkResolver || defaultLinkResolver,
    getProviderToken,
    storeProviderToken,
    storeCallbackState,
    getCallbackState,
    clearCallbackState,
    notifySuccess: props.notifySuccess || notifySuccess,
    settings: {
      renderFooter: props.renderFooter,
    },
    navigate: {
      internal: (page: PageRoute, query?: any) => {
        const u = formatURL(page, query);

        history.push(u);
      },
      external: (url) => {
        if (process.env.NODE_ENV === "test") {
          return;
        }
        window.location.href = url;
      },
    },
    request: window.fetch,
  };

  return <AppContext.Provider {...props} value={value} />;
}
