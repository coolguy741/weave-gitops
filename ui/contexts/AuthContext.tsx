import * as React from "react";
import { useHistory, Redirect } from "react-router-dom";
import { FeatureFlags } from "../contexts/FeatureFlags";

const USER_INFO = "/oauth2/userinfo";
const SIGN_IN = "/oauth2/sign_in";
const LOG_OUT = "/oauth2/logout";
const AUTH_PATH_SIGNIN = "/sign_in";

interface AuthCheckProps {
  children: any;
  Loader?: React.ElementType;
}

export const AuthCheck = ({ children, Loader }: AuthCheckProps) => {
  // If the auth flag is null go straight to rendering the children
  const { authFlag } = React.useContext(FeatureFlags);

  if (!authFlag) {
    return children;
  }

  const { loading, userInfo } = React.useContext(Auth);

  // Wait until userInfo is loaded before showing signin or app content
  if (loading) {
    return Loader ? <Loader /> : null;
  }

  // Signed in! Show app
  if (userInfo?.email) {
    return children;
  }

  // User appears not be logged in, off to signin
  return <Redirect to={AUTH_PATH_SIGNIN} />;
};

export type AuthContext = {
  signIn: (data: any) => void;
  userInfo: {
    email: string;
    groups: string[];
  };
  error: { status: number; statusText: string };
  loading: boolean;
  logOut: () => void;
};

export const Auth = React.createContext<AuthContext | null>(null);

export default function AuthContextProvider({ children }) {
  const { authFlag } = React.useContext(FeatureFlags);
  const [userInfo, setUserInfo] =
    React.useState<{
      email: string;
      groups: string[];
    }>(null);
  const [loading, setLoading] = React.useState<boolean>(true);
  const [error, setError] = React.useState(null);
  const history = useHistory();

  const signIn = React.useCallback((data) => {
    setLoading(true);
    fetch(SIGN_IN, {
      method: "POST",
      body: JSON.stringify(data),
    })
      .then((response) => {
        if (response.status !== 200) {
          setError(response);
          return;
        }
        getUserInfo().then(() => history.push("/"));
      })
      .finally(() => setLoading(false));
  }, []);

  const getUserInfo = React.useCallback(() => {
    setLoading(true);
    return fetch(USER_INFO)
      .then((response) => {
        if (response.status === 400 || response.status === 401) {
          setUserInfo(null);
          return;
        }
        return response.json();
      })
      .then((data) => setUserInfo({ email: data?.email, groups: [] }))
      .catch((err) => console.log(err))
      .finally(() => setLoading(false));
  }, []);

  const logOut = React.useCallback(() => {
    setLoading(true);
    fetch(LOG_OUT, {
      method: "POST",
    })
      .then((response) => {
        if (response.status !== 200) {
          setError(response);
          return;
        }
        history.push("/sign_in");
      })
      .finally(() => setLoading(false));
  }, []);

  React.useEffect(() => {
    if (!authFlag) {
      return null;
    }
    getUserInfo();
    return history.listen(getUserInfo);
  }, [getUserInfo, history]);

  return (
    <>
      {authFlag ? (
        <Auth.Provider
          value={{
            signIn,
            userInfo,
            error,
            loading,
            logOut,
          }}
        >
          {children}
        </Auth.Provider>
      ) : (
        children
      )}
    </>
  );
}
