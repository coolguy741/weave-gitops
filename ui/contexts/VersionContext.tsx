import * as React from "react";
import { AppContext } from "./AppContext";

type Props = {
  children?: any;
};

export type VersionType = {
  version: string;
  gitCommit: string;
  branch: string;
  buildTime: string;
};

export type VersionContextType = {
  data?: VersionType;
  loading: boolean;
  error: Error;
};

export const VersionContext =
  React.createContext<VersionContextType | null>(null);

export default function VersionContextProvider({ children }: Props) {
  const { request } = React.useContext(AppContext);
  const [data, setData] = React.useState(null);
  const [loading, setLoading] = React.useState<boolean>(true);
  const [error, setError] = React.useState(null);

  React.useEffect(() => {
    setLoading(true);
    request("/v1/version")
      .then((response) => response.json())
      .then((data) => setData(data))
      .catch((error) => setError(error))
      .finally(() => setLoading(false));
  }, []);

  const version = data?.version;
  const value: VersionType = version
    &&
    {
      version: version.version,
      gitCommit: version["git-commit"],
      branch: version.branch,
      buildTime: version["buildtime"]
    };

  return (
    <VersionContext.Provider value={{ data: value, loading, error }}>
      {children}
    </VersionContext.Provider>
  );
}
