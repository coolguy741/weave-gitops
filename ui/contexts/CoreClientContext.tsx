import qs from "query-string";
import * as React from "react";
import { Core } from "../lib/api/core/core.pb";
import { AuthRoutes } from "./AuthContext";

type Props = {
  api: typeof Core;
  children: any;
};

export type CoreClientContextType = {
  api: typeof Core;
};

export const CoreClientContext =
  React.createContext<CoreClientContextType | null>(null);

export function UnAuthorizedInterceptor(api: any) {
  const wrapped = {} as any;
  //   Wrap each API method in a check that redirects to the signin page if a 401 is returned.
  for (const method of Object.getOwnPropertyNames(api)) {
    if (typeof api[method] != "function") {
      continue;
    }
    wrapped[method] = (req, initReq) => {
      return api[method](req, initReq).catch((err) => {
        if (err.code === 401) {
          window.location.replace(
            AuthRoutes.AUTH_PATH_SIGNIN +
              "?" +
              qs.stringify({
                redirect: location.pathname + location.search,
              })
          );
        }
        throw err;
      });
    };
  }
  return wrapped;
}

export default function CoreClientContextProvider({ api, children }: Props) {
  const wrapped = UnAuthorizedInterceptor(api) as typeof Core;

  return (
    <CoreClientContext.Provider value={{ api: wrapped }}>
      {children}
    </CoreClientContext.Provider>
  );
}
