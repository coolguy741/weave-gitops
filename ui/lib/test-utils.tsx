import { MuiThemeProvider } from "@material-ui/core";
import { createMemoryHistory } from "history";
import _ from "lodash";
import * as React from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { Router } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import AppContextProvider, { AppProps } from "../contexts/AppContext";
import { CoreClientContext } from "../contexts/CoreClientContext";
import {
  Applications,
  GetGithubAuthStatusRequest,
  GetGithubAuthStatusResponse,
  GetGithubDeviceCodeRequest,
  GetGithubDeviceCodeResponse,
  ParseRepoURLRequest,
  ParseRepoURLResponse,
  ValidateProviderTokenRequest,
  ValidateProviderTokenResponse,
} from "./api/applications/applications.pb";
import {
  Core,
  GetChildObjectsRequest,
  GetChildObjectsResponse,
  GetReconciledObjectsRequest,
  GetReconciledObjectsResponse,
  GetVersionRequest,
  GetVersionResponse,
  ListObjectsRequest,
  ListObjectsResponse,
} from "./api/core/core.pb";
import theme, { muiTheme } from "./theme";
import { RequestError } from "./types";

export type CoreOverrides = {
  GetChildObjects?: (req: GetChildObjectsRequest) => GetChildObjectsResponse;
  GetReconciledObjects?: (
    req: GetReconciledObjectsRequest
  ) => GetReconciledObjectsResponse;
  GetVersion?: (req: GetVersionRequest) => GetVersionResponse;
  ListObjects?: (req: ListObjectsRequest) => ListObjectsResponse;
};

export const createCoreMockClient = (
  ovr: CoreOverrides,
  error?: RequestError
): typeof Core => {
  const promisified = _.reduce(
    ovr,
    (result, handlerFn, method) => {
      result[method] = (req) => {
        if (error) {
          return new Promise((_, reject) => reject(error));
        }
        return new Promise((accept) => accept(handlerFn(req) as any));
      };

      return result;
    },
    {}
  );

  return promisified as typeof Core;
};

export type ApplicationOverrides = {
  GetGithubDeviceCode?: (
    req: GetGithubDeviceCodeRequest
  ) => GetGithubDeviceCodeResponse;
  GetGithubAuthStatus?: (
    req: GetGithubAuthStatusRequest
  ) => GetGithubAuthStatusResponse;
  ParseRepoURL?: (req: ParseRepoURLRequest) => ParseRepoURLResponse;
  ValidateProviderToken?: (
    req: ValidateProviderTokenRequest
  ) => ValidateProviderTokenResponse;
};

// Don't make the user wire up all the promise stuff to be interface-compliant
export const createMockClient = (
  ovr: ApplicationOverrides,
  error?: RequestError
): typeof Applications => {
  const promisified = _.reduce(
    ovr,
    (result, handlerFn, method) => {
      result[method] = (req) => {
        if (error) {
          return new Promise((_, reject) => reject(error));
        }
        return new Promise((accept) => accept(handlerFn(req) as any));
      };

      return result;
    },
    {}
  );

  return promisified as typeof Applications;
};

export function withTheme(element) {
  return (
    <MuiThemeProvider theme={muiTheme}>
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MuiThemeProvider>
  );
}

type TestContextProps = AppProps & { api?: typeof Core };

export function withContext(
  TestComponent,
  url: string,
  { api, ...appProps }: TestContextProps
) {
  const history = createMemoryHistory();
  history.push(url);
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  const isElement = React.isValidElement(TestComponent);
  return (
    <Router history={history}>
      <AppContextProvider renderFooter {...appProps}>
        <QueryClientProvider client={queryClient}>
          <CoreClientContext.Provider value={{ api, featureFlags: {} }}>
            {isElement ? TestComponent : <TestComponent />}
          </CoreClientContext.Provider>
        </QueryClientProvider>
      </AppContextProvider>
    </Router>
  );
}
