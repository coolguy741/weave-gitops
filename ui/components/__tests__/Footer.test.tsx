import { render, screen } from "@testing-library/react";
import "jest-canvas-mock";
import "jest-styled-components";
import React from "react";
import { act } from "react-dom/test-utils";
import { CoreClientContext } from "../../contexts/CoreClientContext";
import {
  createCoreMockClient,
  withContext,
  withTheme,
} from "../../lib/test-utils";
import Footer from "../Footer";

describe("Footer", () => {
  let container;
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });
  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  describe("snapshots", () => {
    it("default", async () => {
      await act(async () => {
        render(
          withTheme(
            withContext(
              <CoreClientContext.Provider
                value={{
                  api: createCoreMockClient({
                    GetVersion: () => ({
                      version: {
                        version: "v0.0.1",
                        branch: "mybranch",
                        "git-commit": "123abcd",
                      },
                    }),
                  }),
                }}
              >
                <Footer />
              </CoreClientContext.Provider>,
              "/",
              {}
            )
          ),
          container
        );
      });

      const footer = screen.getByRole("footer");
      expect(footer).toMatchSnapshot();
    });
    it("no api version", async () => {
      await act(async () => {
        render(
          withTheme(
            withContext(
              <CoreClientContext.Provider
                value={{
                  api: createCoreMockClient({
                    GetVersion: () => ({}),
                  }),
                }}
              >
                <Footer />
              </CoreClientContext.Provider>,
              "/",
              {}
            )
          ),
          container
        );
      });

      const footer = screen.getByRole("footer");
      expect(footer).toMatchSnapshot();
    });
  });
});
