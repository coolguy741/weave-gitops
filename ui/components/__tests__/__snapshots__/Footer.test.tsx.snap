// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer snapshots default 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c3 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 16px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
  color: #737373;
}

.c6 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 16px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
  color: #00b3ec;
}

.c5 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5.title-bar-button {
  width: 250px;
}

.c4 {
  padding: 4px;
}

.c7 {
  padding-right: 24px;
}

.c1 {
  color: #737373;
}

<footer
  className="c0 c1 Footer"
>
  <div
    className="c2"
  >
    <span
      className="c3"
      color="neutral30"
      size="normal"
    >
      Need help? Contact us at
    </span>
    <div
      className="c4"
    />
    <a
      className="c5"
      href="mailto:support@weave.works"
      rel="noreferrer"
      target="_blank"
    >
      <span
        className="c6"
        color="primary"
        size="normal"
      >
        support@weave.works
      </span>
    </a>
  </div>
  <div
    className="c2 c7"
  >
    <div
      className="c4"
    />
    <span
      className="c3"
      color="neutral30"
      size="normal"
    >
      Â© 2022 Weaveworks
    </span>
  </div>
</footer>
`;
